File: python-ruby-study/override/2.py
class Cal(object):
	_history = []
	def __init__(self, v1, v2):
		if isinstance(v1, int):
			self.v1 = v1
		if isinstance(v2, int):
			self.v2 = v2

	def add(self):
		result = self.v1 + self.v2
		Cal._history.append("add : %d+%d=%d" % (self.v1, self.v2, result))
		return result

	def subtract(self):
		result = self.v1 - self.v2
		Cal._history.append("subtract : %d-%d=%d" % (self.v1, self.v2, result))
		return result

	def setV1(self, v):
		if isinstance(v, int):
			self.v1 = v
	def getV1(self):
		return self.v1
	@classmethod
	def history(cls):
		for item in Cal._history:
			print(item)
	def __str__(self):
		return "Cal => v1 : %d, v2 : %d" % (self.v1, self.v2)

class CalMultiply(Cal):
	def multiply(self):
		result = self.v1 * self.v2
		Cal._history.append("multiply : %d*%d=%d" % (self.v1, self.v2, result))
		return result
	def __str__(self):
		return "CalMultiply => %s" % (super(CalMultiply, self).__str__())

class CalDivide(CalMultiply):
	def divide(self):
		result = self.v1 / self.v2
		Cal._history.append("divide : %d/%d=%d" % (self.v1, self.v2, result))
		return result
	def __str__(self):
		return "%s => %s" % (type(self).__name__, super(CalDivide, self).__str__())

p0 = CalDivide(100, 20)
print(p0.divide())
print(p0.add())
print(p0._history)
print(p0)
Cal.history()

File: python-ruby-study/override/2.rb
class Cal
  @@history = []
  def initialize(v1, v2)
    @v1 = v1
    @v2 = v2
  end
  def add
  	result = @v1 + @v2
  	@@history.push("#{@v1}+#{@v2}=#{result}")
  	result
  end
  def subtract
  	result = @v1 - @v2
  	@@history.push("#{@v1}-#{@v2}=#{result}")
  	result
  end
  def setV1(v)
  	if v.is_a?(Integer)
  		@v1 = v
  	end
  end
  
  def inspect
  	"Cal => v1 : #{@v1}, v2 : #{@v2}"
  end

  def Cal.history
  	p "=== history ==="
  	for item in @@history
  		p item
  	end
  	nil
  end

end

class CalMultiply < Cal
  def multiply
  	result = @v1 * @v2
  	@@history.push("#{@v1}*#{@v2}=#{result}")
  	result
  end
  def inspect
  	"CalMultiply => #{super()}"
  end
end

class CalDivide < CalMultiply
  def divide
  	result = @v1 / @v2
  	@@history.push("#{@v1}/#{@v2}=#{result}")
  	result
  end
  def inspect
  	"CalDivide => #{super()}"
  end
end

c = CalDivide.new(100, 40)
p c.multiply
p c.divide
p c.add
p c.subtract
p c
Cal.history